src/activity_item_handling.cpp

```diff
@@ -1159,12 +1153,566 @@ void activity_on_turn_blueprint_move( player_activity &, player &p )
+            const auto items_at = g->m.i_at( p_loc );
+            for( const item &elem : items_at ) {
+                if( elem.has_quality( qual ) ) {
+                    found_required = true;
+                    src_loc = p_loc;
+                    break;
+                }
+            }
+            if( found_required ) {
+                break;
+            }
+        }
+    }
+    if( !found_required ) {
+        // didnt find the required item anywhere, cancel the whole thing
+        return;
+    }
+    if( !g->m.inbounds( src_loc ) ) {
+        if( !g->m.inbounds( p->pos() ) ) {
+            // p is implicitly an NPC that has been moved off the map, so reset the activity
+            // and unload them
+            p->assign_activity( act_multiple_fetch );
+            p->set_moves( 0 );
+            g->reload_npcs();
+            return;
+        }
+        std::vector<tripoint> route = route_adjacent( *p, src_loc );
+        if( route.empty() ) {
+            // can't get there, can't do anything, skip it
+            return;
+        }
+        p->set_destination( route, player_activity( act_multiple_fetch ) );
+        return;
+    }
+
+    // dont go there if it's dangerous.
+    if( g->m.dangerous_field_at( src_loc ) ) {
+        return;
+    }
+    bool adjacent = false;
+    for( auto elem : g->m.points_in_radius( src_loc, 1 ) ) {
+        if( p->pos() == elem ) {
+            adjacent = true;
+            break;
+        }
+    }
+    if( !adjacent ) {
+        std::vector<tripoint> route = route_adjacent( *p, src_loc );
+
+        // check if we found path to source / adjacent tile
+        if( route.empty() ) {
+            add_msg( m_info, _( "%s can't reach the source tile to fetch required items" ),
+                     p->disp_name() );
+            return;
+        }
+        if( p->moves <= 0 ) {
+            // Restart activity and break from cycle.
+            p->assign_activity( act_multiple_fetch );
+            return;
+        }
+        // set the destination and restart activity after player arrives there
+        // we don't need to check for safe mode,
+        // activity will be restarted only if
+        // player arrives on destination tile
+        p->set_destination( route, player_activity( act_multiple_fetch ) );
+        return;
+    }
+
+    // find the tool we spotted earlier, now we are nearby.
+    // Maybe its moved in the meantime.
+    auto items_at_post = g->m.i_at( src_loc );
+    bool found_post = false;
+    for( map_stack::iterator it = items_at_post.begin(); it != items_at_post.end(); ) {
+        if( it->has_quality( qual ) ) {
+            units::volume volume_allowed = p->volume_capacity() - p->volume_carried();
+            units::mass weight_allowed = p->weight_capacity() - p->weight_carried();
+            if( volume_allowed <= 0_ml || weight_allowed <= 0_gram ) {
+                add_msg( m_info, _( "%s cannot pick up the %s, it's too big." ), p->disp_name(), it->tname() );
```

Comment:
Yeah, infinite looping. At the end of this function, the activity is restarted (comment indicates this is because the item has gone missing), but it also happens when the NPC can not pick this item up. Next time, the NPC sees the very same item again and attempts to pick it up again, only to detect it won't fit. And the cycle begins again.


Why is the NPC actually going to that item in the first place? They should already know it can not be picked up!